File: Java/easy/cousinsInBT.java
import java.util.*;

class Solution {
    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    private static class TestHolder{
        TreeNode root;
        int x, y;
        TestHolder(TreeNode root, int x, int y){
            this.root = root;
            this.x = x;
            this.y = y;
        }
    }
    private static TreeNode xParent = null, yParent = null;
    private static int xDepth = -1, yDepth = -1;

    private static void findVal(TreeNode node, TreeNode parent, int x, int y, int currentDepth){
        if(node == null)    return;
        if(yParent != null && xParent != null)  return;
        findVal(node.left, node, x, y, currentDepth + 1);
        if(node.val == x){
            xDepth = currentDepth;
            xParent = parent;
        }
        if(node.val == y){
            yDepth = currentDepth;
            yParent = parent;
        }
        if(yParent != null && xParent != null)  return;
        findVal(node.right, node, x, y, currentDepth + 1);
    }

    public static boolean isCousins(TreeNode root, int x, int y) {
        findVal(root, root, x, y, 0);
        return xDepth == yDepth && xParent != yParent;
    }

    public static void inorder(TreeNode node){
        if(node == null)    return;
        inorder(node.left);
        System.out.print(" " + node.val);
        inorder(node.right);
    }

    // Tree = 4 2 1 3
    // x = 4 and y = 3 are not cousins
    
    // Tree = 2 4 1 3 5
    // x = 5 and y = 4 are cousins
    
    // Tree = 2 4 1 3
    // x = 2 and y = 3 are not cousins

    public static void main(String[] args){
        TreeNode root1 = new TreeNode(1, new TreeNode(2, new TreeNode(4),null), new TreeNode(3));
        TreeNode root2 = new TreeNode(1, new TreeNode(2, null, new TreeNode(4)), new TreeNode(3, null, new TreeNode(5)));
        TreeNode root3 = new TreeNode(1, new TreeNode(2, null, new TreeNode(4)), new TreeNode(3));

        TestHolder[] testSet = new TestHolder[]{
            new TestHolder(root1, 4, 3),
            new TestHolder(root2, 5, 4),
            new TestHolder(root3, 2, 3)
        };

        for(TestHolder test : testSet){
            xDepth = -1;
            yDepth = -1;
            xParent = null;
            yParent = null;
            System.out.print("Tree =");
            inorder(test.root);
            System.out.println("\nx = " + test.x + " and y = " + test.y + " are" + (isCousins(test.root, test.x, test.y)? "":" not") + " cousins\n");
        }
    }
}

File: CPP/easy/cousinsInBT.cpp
#include <vector>
#include <iostream>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
private:
    int xDepth = 0, yDepth = 0;
    TreeNode* xParent = nullptr, *yParent = nullptr;

public:
    void found(int x, int y, TreeNode* node, TreeNode* parent, int currentDepth){
        if(!node)   return;
        if(xParent && yParent)  return;
        found(x, y, node->left, node, currentDepth + 1);
        if(node->val == x){
            xParent = parent;
            xDepth = currentDepth;
        }
        if(node->val == y){
            yParent = parent;
            yDepth = currentDepth;
        }
        found(x, y, node->right, node, currentDepth + 1);
        if(xParent && yParent)  return;
    }

    bool isCousins(TreeNode* root, int x, int y) {
        found(x, y, root, root, 0);
        return xParent != yParent && xDepth == yDepth;
    }
};

void inorder(TreeNode* node){
    if(!node)   return;
    inorder(node->left);
    std::cout << " " << node->val ;
    inorder(node->right); 
}

struct TestHolder{
    TreeNode* root;
    int x, y;
    TestHolder(TreeNode* r, int xVal, int yVal) : root(r), x(xVal), y(yVal){}
};

// Tree = 4 2 1 3
// x = 4 and y = 3 are not cousins

// Tree = 2 4 1 3 5
// x = 5 and y = 4 are cousins

// Tree = 2 4 1 3
// x = 2 and y = 3 are not cousins

int main(){
    TreeNode* root1 = new TreeNode(1, new TreeNode(2, new TreeNode(4),nullptr), new TreeNode(3));
    TreeNode* root2 = new TreeNode(1, new TreeNode(2, nullptr, new TreeNode(4)), new TreeNode(3, nullptr, new TreeNode(5)));
    TreeNode* root3 = new TreeNode(1, new TreeNode(2, nullptr, new TreeNode(4)), new TreeNode(3));

    std::vector<TestHolder*> testSet = {
        new TestHolder(root1, 4, 3),
        new TestHolder(root2, 5, 4),
        new TestHolder(root3, 2, 3)
    };

    for(TestHolder* test : testSet){
        Solution* sol = new Solution();
        std::cout << "Tree =";
        inorder(test->root);
        std::cout << "\nx = " << test->x << " and y = " << test->y << " are" << (sol->isCousins(test->root, test->x, test->y)? "":" not") << " cousins\n" << std::endl;
    }
    return 0;
}